lib/src/kbasesearchengine/system/IndexingRules.java

```diff
@@ -2,23 +2,205 @@
 
 import kbasesearchengine.common.ObjectJsonPath;
 
+/**
+ * This class defines the rules for parsing source data, collecting portions
+ * of it, and preparing it for indexing. A single instance of this class would
+ * declare the indexing rules for a single keyword within a source object or
+ * sub-object. A source object or sub-object would typically require many
+ * indexing rules for the set of keywords that would need to be indexed for the
+ * source (sub-)object.
+ *
+ * Each indexing rule forms a key-value pair based on some
+ * content of the object (or sub-object) that it is to be indexed by the rules.
+ * An indexing rule may also be based on a key-value pair formed by another
+ * indexing rule.
+ *
+ * See {@link #validate()} for instructions on building unambigious indexing rules.
+ *
+ */
 public class IndexingRules {
```

Comment:
I'm not saying do this now, but at some point we should move to a builder for an immutable class so the validate happens on a build.